### DISCUSSION TITLE
METIS vs Rtree (partitioning)

### DISCUSSION THREAD

**Discussion by fdrmrc** [fdrmrc/non_matching_test_suite#35] [Show and tell]
@luca-heltai @andreacangiani 

I will collect on this discussion some other examples as the ones of this morning. Feel free to ask for other geometries as well.

- - - - - - - - - - - -
## EXAMPLE 1: Unstructured ~~square~~ rectangular mesh
- Size underlying fine tria: 21 760 elements:

- Number of partitions: **22**
Left: _METIS_
Right: _R-tree_
![22unstructured](https://github.com/fdrmrc/Polydeal/assets/50499769/626c3c8d-834a-4b32-93be-e9aeb866c17d)

- - - - - - - - - - - -

- Number of partitions: **85**
Left: _METIS_
Right: _R-tree_
![85unstructured](https://github.com/fdrmrc/Polydeal/assets/50499769/23c4fb00-bc75-4c64-a2b0-d1ad869f397c)

- - - - - - - - - - - -

- Number of partitions: **340**
Left: _METIS_ 
Right: _R-tree_
![340unstructured](https://github.com/fdrmrc/Polydeal/assets/50499769/dffbc7e6-6e70-4296-a43b-778872bfa3e2)

- - - - - - - - - - - -

**Comment by andreacangiani** [comment]
This is NOT good....it IS FANTASTIC!!! 

ps: what is the domain? It looks rectangular...

**Reply by fdrmrc** [reply]
Yes, the domain here is $[0,1.25] \times [0,1]$!

**Reply by andreacangiani** [reply]
There is  a consistent bias with fewer elements on the right-hand side. It seems to me that this is due to the fact that the domain cannot be divided evenly in the number of agglomerates you fix. Is the underlying fine mesh uniform? There is something weird happening in the top-right corner: for 22 agglomerates the element in the top-right corner is clearly much smaller than the one on its left. Instead, in 85 case, it is bigger. My understanding was that the number of leaves should be the same for every agglomerate, no?

**Reply by luca-heltai** [reply]
As far as I understand, if a leaf has fewer elements, then its parent is allowed to have (1) more element, so that balancing is maintained almost everywhere. This could be the you are observing. The underlying grid is unstructured. Maybe the size of the elements is different on the corner. Marco?

**Reply by fdrmrc** [reply]
Sorry for replying now, yesterday I was at home sick in bed. 

Here's the underlying grid, the size of the elements on the top right corner looks the same as the other elements.
![288996769-7616eecb-7bfa-4cfa-8b33-4341895f3ee2](https://github.com/fdrmrc/Polydeal/assets/50499769/48282408-e86e-4989-8c22-6d81f67d4d2e)

**Reply by andreacangiani** [reply]
Then I don't understand why the agglomerate are so uneven. Perhaps we should look at the whole tree as Luca suggeste, but, to start with, can you plot the list of the numbers of leaves each agglomerate is made of? 
(I am in bed myself now with a nasty cold!)

**Reply by fdrmrc** [reply]
> can you plot the list of the numbers of leaves each agglomerate is made of?

For 22 agglomerates we have:
```
N subdomains = 22
1024 cells have subdomain 0
1024 cells have subdomain 1
1024 cells have subdomain 2
1024 cells have subdomain 3
1024 cells have subdomain 4
1024 cells have subdomain 5
1024 cells have subdomain 6
1024 cells have subdomain 7
1024 cells have subdomain 8
1024 cells have subdomain 9
1024 cells have subdomain 10
1024 cells have subdomain 11
1024 cells have subdomain 12
1024 cells have subdomain 13
1024 cells have subdomain 14
1024 cells have subdomain 15
1024 cells have subdomain 16
1024 cells have subdomain 17
1024 cells have subdomain 18
1024 cells have subdomain 19
1024 cells have subdomain 20
256 cells have subdomain 21
```

**Reply by andreacangiani** [reply]
Interesting....so indeed it maintains perfectly the subdivision but then the last agglomerate is made of the remaining ones. That's not ideal. Mumble, mumble.

**Reply by fdrmrc** [reply]
In one of the following messages I played with the parameters of the rtree (see here https://github.com/fdrmrc/Polydeal/discussions/48#discussioncomment-7798830) . The situation looks better to me in that instance

**Comment by fdrmrc** [comment]
If I take the same domain (the rectangle $[0,1.25] \times [0,1]$) , **but structured**, things are again okay with the tree.


- N partitions: 16
![structured16](https://github.com/fdrmrc/Polydeal/assets/50499769/72c5b5ea-1f5d-455d-aaae-205c09a6143d)

- N partitions: 64
![structured64](https://github.com/fdrmrc/Polydeal/assets/50499769/0cd539c1-6ddf-4993-8334-3ca2a7d23638)

- N partitions: 256
![structured256](https://github.com/fdrmrc/Polydeal/assets/50499769/fdd8ac1b-e29a-49ac-b544-dabba3ad72aa)

- N partitions: 1024
![structured1024](https://github.com/fdrmrc/Polydeal/assets/50499769/6e166b5e-e991-45f4-a5e5-ab4404389793)

**Comment by fdrmrc** [comment]
I played with the parameters of the rtree. The grids above have been obtained with a _maximum number of entries per node = 4._ Setting this parameter to 2 we have better(?) results. In the following I am comparing the Rtree results after changing this value. For the sake of comparison, on the left column I am displaying the agglomerates we obtained before.

**LEFT: max entries per node = 4** ---------- **RIGHT: max entries per node = 2**

- N partitions = 22
![rtree_template_comparison_22](https://github.com/fdrmrc/Polydeal/assets/50499769/7d4456a6-d359-4d34-a597-b71c76e4a66b)

- N partitions = 85
![rtree_template_comparison_85](https://github.com/fdrmrc/Polydeal/assets/50499769/2ed46a2b-1b8e-4a41-89c5-382df399217c)



- N partitions = 340
![rtree_template_comparison_340](https://github.com/fdrmrc/Polydeal/assets/50499769/8b6b6e51-9845-497d-8a05-cd810ed9c9a1)
